### PR TITLE
Add field path tracking for nested structures

### PR DESCRIPTION
**Pull request by t4kamura**
## Description

As part of the resolution for issue #86 (handling path names for nested structures in errors):
By propagating the path of the target fields to the validator, we now provide the correct field path in the error details.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)

## Quality Checklist

Before submitting this pull request, ensure your contribution meets these quality standards:

### Code Quality
- [x] Code passes `make golangci-lint` without errors
- [x] All tests pass: `make test`
- [x] Code follows existing patterns and conventions
- [x] No hard-coded values (use constants or configuration)
- [x] Error messages are clear and actionable

### Testing Requirements
- [x] Golden tests created and passing (`internal/analyzers/govalid/testdata/`)
- [x] Unit tests implemented with boundary value testing (`test/unit/`)
- [ ] Benchmark tests comparing against popular validation libraries (`test/benchmark/`)
  - [ ] [go-playground/validator](https://github.com/go-playground/validator) (BenchmarkGoPlayground*)
  - [ ] [asaskevich/govalidator](https://github.com/asaskevich/govalidator) (BenchmarkGoValidator*)
  - [ ] [go-ozzo/ozzo-validation](https://github.com/go-ozzo/ozzo-validation) (BenchmarkOzzoValidation*)
  - [ ] [gookit/validate](https://github.com/gookit/validate) (BenchmarkGookitValidate*)
- [x] All test files generated: `cd test && go generate`

### Performance Standards
- [ ] Zero allocations in validation logic (verified via benchmarks)
- [ ] Performance improvement over existing validation libraries (minimum 2x faster)
- [ ] Benchmark results added to `test/benchmark/README.md`

### Documentation
- [ ] README.md updated with new marker documentation
- [x] Code includes appropriate comments and examples
- [ ] Validator usage examples provided

### Integration
- [ ] Validator scaffold generated with `make generate-validator MARKER=yourmarker`
- [ ] Registry files automatically updated
- [x] Binary rebuilt and installed: `go install ./cmd/govalid/`

### Final Verification
- [x] All GitHub Actions CI checks pass
- [x] No breaking changes to existing functionality
- [x] Contribution follows project license requirements

## Additional Notes

I've modified the validation generation to propagate the `parentPath`. 

- I'm using this to create `FieldPath` and call it through the `Validator` struct's `FieldPath()` method.
- When generating each error, I inject the appropriate path using `FieldPath`.

As mentioned in [#86](https://github.com/sivchari/govalid/issues/86#issuecomment-3148476131), I've limited the changes to paths for nested struct structures only.

I've also updated the error variable names to include field path information. For example:

```go
type LTE struct {
	Struct struct {
		// +govalid:lte=50
		Value int
	}
}
```
Old: `ErrLTEValueLTEValidation`
New: `ErrLTEStructValueLTEValidation`

This change is necessary to prevent duplicate error variables when field path information is not included in the error variable names.

`internal/validator/fieldpath.go`:
I needed to handle field paths with and without dots, so I've made the field paths more versatile by creating a FieldPath structure. If you see any potential improvements to this approach, please review.

## Related Issues

#86

### CODE REVIEW

File: internal/analyzers/govalid/testdata/src/alpha/govalid.golden [hunk 1]
```diff
@@ -12,13 +12,13 @@ var (
...
-	ErrAlphaFirstNameAlphaValidation = govaliderrors.ValidationError{Reason: "field FirstName must be alphabetic", Path: "Alpha.AlphaFirstName", Type: "alpha"}
+	ErrAlphaFirstNameAlphaValidation = govaliderrors.ValidationError{Reason: "field FirstName must be alphabetic", Path: "Alpha.FirstName", Type: "alpha"}
 
 	// ErrAlphaLastNameAlphaValidation is the error returned when field LastName is not alphabetic.
-	ErrAlphaLastNameAlphaValidation = govaliderrors.ValidationError{Reason: "field LastName must be alphabetic", Path: "Alpha.AlphaLastName", Type: "alpha"}
+	ErrAlphaLastNameAlphaValidation = govaliderrors.ValidationError{Reason: "field LastName must be alphabetic", Path: "Alpha.LastName", Type: "alpha"}
 
 	// ErrAlphaCountryCodeAlphaValidation is the error returned when field CountryCode is not alphabetic.
-	ErrAlphaCountryCodeAlphaValidation = govaliderrors.ValidationError{Reason: "field CountryCode must be alphabetic", Path: "Alpha.AlphaCountryCode", Type: "alpha"}
+	ErrAlphaCountryCodeAlphaValidation = govaliderrors.ValidationError{Reason: "field CountryCode must be alphabetic", Path: "Alpha.CountryCode", Type: "alpha"}
```
[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: @BlackBuck 
I don't manually modify golden files.
> I don't understand why, only for this file, are the errors in the format Err[StructName][FieldName][MarkerName]Validation and others have it as Err[StructName]Struct[FieldName][MarkerName]Validation.

This is because in HEAD, only alpha's Err() method is `fmt.Sprintf(“%s.%s”, v.structName, v.structName+v.FieldName())`.
Other markers are `fmt.Sprintf(“%s.%s”, m.structName, m.FieldName())`.
https://github.com/sivchari/govalid/blob/8d10ce0f2123e374999e79bb8dc11fdd4f062a9a/internal/validator/rules/alpha.go#L49

File: internal/analyzers/govalid/govalid.go [hunk 2]
```diff
@@ -178,7 +178,7 @@ func analyzeMarker(pass *codegen.Pass, markersInspect markers.Markers, structTyp
 	return analyzed
 }
 
-func makeValidator(pass *codegen.Pass, markers markers.MarkerSet, field *ast.Field, structName string) []validator.Validator {
+func makeValidator(pass *codegen.Pass, markers markers.MarkerSet, field *ast.Field, structName, parentPath string) []validator.Validator {
```
Comment (ccoVeille) [inline internal/analyzers/govalid/govalid.go:181]: At some point, it should be considered to pass a struct as argument here instead of adding argument over and over.

Maybe it's the moment

File: internal/validator/registry/registry.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ type ValidatorInitializer interface {
 }
 
 // ValidatorFactory is a function that creates a validator instance.
-type ValidatorFactory func(pass *codegen.Pass, field *ast.Field, expressions map[string]string, structName string, ruleName string) validator.Validator
+type ValidatorFactory func(pass *codegen.Pass, field *ast.Field, expressions map[string]string, structName, ruleName, parentPath string) validator.Validator
```
Comment (ccoVeille) [inline internal/validator/registry/registry.go:23]: Same remark than here

https://github.com/sivchari/govalid/pull/134#discussion_r2264624173

There is too many fields, using a struct would be better

File: internal/validator/fieldpath.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+		}
+	}
+
+	return FieldPath(strings.Join(nonEmpty, "."))
+}
+
+// WithoutDots returns the field path with all dots removed for use in variable names and keys.
+func (fp FieldPath) WithoutDots() string {
+	return strings.ReplaceAll(string(fp), ".", "")
+}
```
Comment (ccoVeille) [inline internal/validator/fieldpath.go:25]: May I suggest this?

```suggestion
// CleanedPath returns the field path with all dots removed for use in variable names and keys.
func (fp FieldPath) CleanedPath() string {
	return strings.ReplaceAll(string(fp), ".", "")
}
```

File: internal/analyzers/govalid/testdata/src/alpha/govalid.golden (hunk 1 above)
[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: Yeah, but you [changed it](https://github.com/sivchari/govalid/blob/c2c41d10cc9a282146d4e0249b0c0e66dd418308/internal/validator/rules/alpha.go#L56C1-L56C37) so you probably have to reinstall with this PR's current HEAD using `go install cmd/govalid` for it to work

[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: I thought it was a mistake that the alpha golden file in the main branch currently has `fmt.Sprintf(“%s.%s”, v.structName, v.structName+v.FieldName())`.
Does fixing this have any impact?
Or is this difference something that should not occur?

[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: Ideally, this shouldn't be occurring (since the code for `CleanedPath` and `[@PATH]` replacement is the same for all the validators). 
Fixing this means that all errors are consistent and predictable (for better user and debugging experience).

[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: > this shouldn't be occurring

which means

> the errors in the format `Err[StructName][FieldName][MarkerName]Validation` and others have it as `Err[StructName]Struct[FieldName][MarkerName]Validation`.

should not occur?

In this PR, all error variable names are in the following format:
`Err[StructName][FieldName][MarkerName]Validation`

Rather than [StructName] and [FieldName] being fixed as `Struct`, only the names of such nested structures are included.
The current implementation uses consistent error variable names.
Please let me know if I am mistaken.

[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: What I meant to say was that even though the code for `CleanedPath` for all validators is the same, the `Path` isn't reflected in the `Err` variable for  `alpha` validator.

Others:

```go
// ErrGTStructIntGTValidation is the error returned when the value of the field is less than the 1.
	ErrGTStructIntGTValidation = govaliderrors.ValidationError{Reason: "field Int must be greater than 1", Path: "GT.Struct.Int", Type: "gt"}
```

alpha.go

```go
ErrAlphaLastNameAlphaValidation = govaliderrors.ValidationError{Reason: "field LastName must be alphabetic", Path: "Alpha.LastName", Type: "alpha"}
```

Notice the missing `.Struct`?

 It's surprising to see that it's the only one that behaves like this.

**Review by BlackBuck — APPROVED** [review]

File: internal/analyzers/govalid/testdata/src/alpha/govalid.golden (hunk 1 above)
[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: Hey @t4kamura 
I'm really sorry for the slip-up!! It was a mistake on my end. It turns out, I got confused when you mentioned the current HEAD's version. I missed the part where the new nested struct had been added to the `LTE` test and got confused. The changes look good from my end :)

[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/analyzers/govalid/testdata/src/alpha/govalid.golden:21]: No problem, thanks for the review!

**Comment by sivchari** [comment]
Sorry for late response. I'd take a look by tomorrow.

File: internal/analyzers/govalid/testdata/src/lte/govalid.golden [hunk 5]
```diff
@@ -16,8 +16,8 @@ var (
 	// ErrLTEScoreLTEValidation is the error returned when the value of the field is greater than 10.5.
 	ErrLTEScoreLTEValidation = govaliderrors.ValidationError{Reason: "field Score must be less than or equal to 10.5", Path: "LTE.Score", Type: "lte"}
 
-	// ErrLTEValueLTEValidation is the error returned when the value of the field is greater than 50.
-	ErrLTEValueLTEValidation = govaliderrors.ValidationError{Reason: "field Value must be less than or equal to 50", Path: "LTE.Value", Type: "lte"}
+	// ErrLTEStructValueLTEValidation is the error returned when the value of the field is greater than 50.
```
[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden:19]: @sivchari 
Currently, error variables are not always in a 1:1 relationship with specific fields. For example:

```go
type A struct {
	Struct struct {
		// +govalid:required
		Value int
	}
	Struct2 struct {
		// +govalid:required
		Value int
	}
}
// In the current implementation, these two fields are consolidated into a single error variable.
```

I acknowledge that this PR will introduce a breaking change. However, I believe it is necessary to implement a 1:1 relationship between fields and their error variables. I am in favor of releasing this as a version 2 (v2) release to accommodate this significant modification.

File: internal/validator/fieldpath_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
+package validator_test
+
+import (
+	"testing"
+
+	"github.com/sivchari/govalid/internal/validator"
+)
+
+func TestNewFieldPath(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (t4kamura) [reply internal/validator/fieldpath_test.go:9]: @sivchari 
Thank you for the suggestion!
Are you suggesting that you want to include the array index number in the error variable? Or do you simply want to include the index number within the `components` stored in the FieldPath?

Currently, since validation is being generated at the field level, incorporating index numbers would require developing a mechanism to generate errors during the validation process.

[the remark this answers is not in the document]
Comment (sivchari) [reply internal/validator/fieldpath_test.go:9]: I expected to include index number in the FieldPath.

> Currently, since validation is being generated at the field level, incorporating index numbers would require developing a mechanism to generate errors during the validation process.

You're right, so please open the issue when we make it feature-request.
Thanks.

File: internal/analyzers/govalid/testdata/src/lte/govalid.golden (hunk 5 above)
[the remark this answers is not in the document]
Comment (sivchari) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden:19]: No, you should use new error variable, then old error variable only has alias to new error variable.

```go
type A struct {
	Struct struct {
		// +govalid:required
		Value int
	}
}
```

```go
var (
	// ErrNilA is returned when the A is nil.
	ErrNilA = errors.New("input A is nil")

	// Deprecated:
	// ErrAValueRequiredValidation is returned when the Value is required but not provided.
	ErrAValueRequiredValidation = ErrAStructValueRequiredValidation

	// ErrAStructValueRequiredValidation is returned when the Value is required but not provided.
	ErrAStructValueRequiredValidation = govaliderrors.ValidationError{Reason: "field Value is required", Path: "A.Struct.Value", Type: "required"}
)

func ValidateA(t *A) error {
	if t == nil {
		return ErrNilA
	}

	var errs govaliderrors.ValidationErrors

	{
		t := t.Struct

		if t.Value == 0 {
			err := ErrAStructValueRequiredValidation
			err.Value = t.Value
			errs = append(errs, err)
		}

	}

	if len(errs) > 0 {
		return errs
	}
	return nil
}
```

**Review by ccoVeille — APPROVED** [review]

File: internal/analyzers/govalid/testdata/src/lte/govalid.golden (hunk 5 above)
[the remark this answers is not in the document]
Comment (sivchari) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden:24]: @t4kamura 
I'm sorry if I miss your changes because of the differences are too big, but did you also add the deprecation syntax to [this file](https://github.com/sivchari/govalid/blob/main/cmd/generate-validators/templates/validator.go.tmpl) in template. We should add these changes to template file.

File: cmd/generate-validators/templates/validator.go.tmpl
Comment (sivchari) [inline cmd/generate-validators/templates/validator.go.tmpl:1]: We should add LEGACYERRVARIABLE to this file.

Comment (t4kamura) [reply cmd/generate-validators/templates/validator.go.tmpl:1]: @sivchari 
Thank you.
I've added legacy errors to the template file, and I kindly request that you review it when you have a moment.

File: cmd/generate-validators/templates/validator.go.tmpl [hunk 7]
```diff
@@ -31,31 +34,51 @@ func (v *{{.MarkerName}}Validator) FieldName() string {
...
 	`
 
+  legacyErrVarName := fmt.Sprintf("Err%s%s{{.StructName}}Validation", v.structName, v.FieldName())
+	currentErrVarName := v.ErrVariable()
+
 	replacer := strings.NewReplacer(
-		"[@ERRVARIABLE]", v.ErrVariable(),
-		"[@FIELD]", v.structName+v.FieldName(),
-		"[@PATH]", fmt.Sprintf("%s.%s", v.structName+v.FieldName()),
+    "[@ERRVARIABLE]", currentErrVarName,
```
Comment (sivchari) [inline cmd/generate-validators/templates/validator.go.tmpl:66]:
```suggestion
        "[@ERRVARIABLE]", currentErrVarName,
```

Comment (sivchari) [reply cmd/generate-validators/templates/validator.go.tmpl:66]: Please align indent.
